src/ds/address.h

```diff
@@ -39,4 +39,33 @@ namespace snmalloc
   {
     return reinterpret_cast<T*>(address);
   }
+
+  template<size_t granule, typename T = void>
+  inline T* pointer_align_down(void* p)
+  {
+    static_assert(granule > 0);
+    static_assert((granule & (granule - 1)) == 0);
+#if defined(__has_builtin) && __has_builtin(__builtin_align_down)
+    return reinterpret_cast<T*>(__builtin_align_down(p, granule));
+#else
+    /* XREF bits::align_down; can't use here for cyclic deps */
+    size_t align_1 = granule - 1;
+    return reinterpret_cast<T*>(reinterpret_cast<uintptr_t>(p) & ~align_1);
+#endif
+  }
+
+  template<size_t granule, typename T = void>
+  inline T* pointer_align_up(void* p)
+  {
+    static_assert(granule > 0);
+    static_assert((granule & (granule - 1)) == 0);
+#if defined(__has_builtin) && __has_builtin(__builtin_align_up)
+    return reinterpret_cast<T*>(__builtin_align_up(p, granule));
+#else
+    /* XREF bits::align_up; can't use here for cyclic deps */
+    size_t align_1 = granule - 1;
+    return reinterpret_cast<T*>(
+      reinterpret_cast<uintptr_t>(pointer_offset(p, granule)) & ~align_1);
```

Comment:
I think the problem with 32bit build is caused by:
```
pointer_offset(p, granule)
```
should be
```
pointer_offset(p, align_1)
```
This is causing a lot of Virtual address space to be wasted, and leading to the issue.

Reply:
Ah ha.  Good catch.  I'm out of the office right now but I'll push a replacement as soon as I can.